containerd-shim-spin/src/engine.rs

```diff
@@ -0,0 +1,238 @@
+
+    fn resolve_app_source(&self, app_source: PathBuf) -> Result<ResolvedAppSource> {
+        Ok(ResolvedAppSource::File {
+            manifest_path: app_source.clone(),
+            manifest: spin_manifest::manifest_from_file(app_source)?,
+        })
+    }
+
+    async fn wasm_exec_async(&self) -> Result<()> {
+        let app_source = self.app_source()?;
+        let resolved_app_source = self.resolve_app_source(app_source.clone())?;
+        let trigger_cmd = trigger_command_for_resolved_app_source(&resolved_app_source)
+            .with_context(|| format!("Couldn't find trigger executor for {app_source:?}"))?;
+        let mut locked_app = self.load_resolved_app_source(resolved_app_source).await?;
+        self.update_locked_app(&mut locked_app); // no-op for now
+        self.run_trigger(&trigger_cmd, locked_app).await
+    }
+
+    async fn run_trigger(&self, trigger_type: &str, app: LockedApp) -> Result<()> {
+        let working_dir = PathBuf::from("/");
+        let f = match trigger_type {
+            HttpTrigger::TRIGGER_TYPE => {
+                let http_trigger: HttpTrigger = self
+                    .build_spin_trigger(working_dir, app)
+                    .await
+                    .context("failed to build spin trigger")?;
+                info!(" >>> running spin trigger");
+                http_trigger.run(spin_trigger_http::CliArgs {
+                    address: parse_addr(SPIN_ADDR).unwrap(),
+                    tls_cert: None,
+                    tls_key: None,
+                })
+            }
+            RedisTrigger::TRIGGER_TYPE => {
+                let redis_trigger: RedisTrigger = self
+                    .build_spin_trigger(working_dir, app)
+                    .await
+                    .context("failed to build spin trigger")?;
+
+                info!(" >>> running spin trigger");
+                redis_trigger.run(spin_trigger::cli::NoArgs)
+            }
+            _ => {
+                todo!("Only Http and Redis triggers are currently supported.")
+            }
+        };
+        info!(" >>> notifying main thread we are about to start");
+        f.await
+    }
+
+    async fn load_resolved_app_source(
+        &self,
+        resolved: ResolvedAppSource,
+    ) -> anyhow::Result<LockedApp> {
+        match resolved {
+            ResolvedAppSource::File { manifest_path, .. } => {
+                // TODO: This should be configurable, see https://github.com/deislabs/containerd-wasm-shims/issues/166
+                let files_mount_strategy = FilesMountStrategy::Direct;
+
+                // create a cache directory at /.cache
+                let cache_dir = PathBuf::from("/.cache");
+                env::set_var("XDG_CACHE_HOME", &cache_dir);
+
+                if !cache_dir.exists() {
+                    tokio::fs::create_dir_all(&cache_dir)
+                        .await
+                        .with_context(|| format!("failed to create {:?}", cache_dir))?;
+                }
```

Comment:
What's the cache for? I don't see this being used anywhere

Reply:
`spin_loader::from_file` uses private `LocalLoder` and in turn uses `Cache::new` and that uses `dirs::cache_dir()` which expects `$HOME/.cache` or `XDG_CACHE_HOME`. We don't have both, so this cache is created for this purpose.

Reply:
Otherwise, the spin shim would just fail with an error message at containerd log saying "cannot get cache directory" which precisely was hitting here: https://github.com/fermyon/spin/blob/61217c2b897334bdc3f9d0c41509534be2e92134/crates/loader/src/cache.rs#L26-L28

Reply:
This is some inconvenience that a spin refactory introduces. We can work with them to figure out a better solution here. @lann FYI https://github.com/fermyon/spin/pull/1780

Reply:
We should document this via comments

Reply:
Documented

Reply:
did we open a issue to track this in spin?

Reply:
https://github.com/fermyon/spin/issues/2072

Reply:
The cache directory is used to store downloaded remote component sources, like https://github.com/fermyon/spin/blob/7ce1ca058f66ffa1e590aa0dd8efa1ea33a6f346/templates/static-fileserver/content/spin.toml#L14

Its helpful to use the common cache for this as it deduplicates these downloads between apps. We can add some option to `LocalLoader` to support whatever the shim wants to do here.

Reply:
For OCI support, the OCI loader takes a cache in explicitly.  https://github.com/fermyon/spin/blob/40bbead627ef6c8f61706516e5f11f1122a33f21/crates/oci/src/loader.rs#L45

Even thought we create it explicitly the `LocalLoader` loads we need the additional ENV set. https://github.com/deislabs/containerd-wasm-shims/pull/180/files#diff-54a5048ac3353fa7e095d007af9b9fa55d60219c1111b9435572283b405f0f23R134-R135

It seem that `LocalLoader`, should have a way to set the `Cache` object so the API's are similiar.